File: GFG/New Year Resolution.py
from typing import List
class Solution:
    def f(self, day, total, N, coins,dp):
        if total != 0 and (total % 24 == 0 or total % 20 == 0 or total == 2024):
            return True
        if day >= N:
            return False
        if dp[day][total] != -1:
            return dp[day][total]
        a = self.f(day + 1, total + coins[day], N, coins,dp)
        b = self.f(day + 1, total, N, coins,dp)
        dp[day][total] = a or b
        return dp[day][total]

    def isPossible(self, N: int, coins: List[int]) -> bool:
        total = 0
        for i in coins:
            total += i
        dp = [[-1 for j in range(total + 1)] for i in range(N)]
        return self.f(0, 0, N, coins,dp)
        

File: GFG/New Year Resolution.cpp
//User function Template for C++

class Solution {
    private:
    bool f(int day,int total,int N,int coins[],vector<vector<int>> &dp) {
        if(total and (!(total % 20) or !(total % 24) or total == 2024)) return true;
        if(day >= N) return false;
        if(dp[day][total] != -1) return dp[day][total];
        int a = f(day + 1,total + coins[day],N,coins,dp);
        int b = f(day + 1,total,N,coins,dp);
        return dp[day][total] = a or b;
    }
    public:
    int isPossible(int N , int coins[]) 
    {   
        int total = 0;
        for(int i = 0; i<N; i++) total += coins[i];
        vector<vector<int>> dp(N,vector<int>(total + 1,-1));
        return f(0,0,N,coins,dp);
    }
    
};